django/core/cache/backends/memcached.py

```diff
@@ -213,3 +213,55 @@ def _modify_servers(servers):
                 yield server[5:]
             else:
                 yield server
+
+
+class PyMemcacheCache(BaseMemcachedCache):
+    """An implementation of a cache binding using pymemcache."""
+
+    def __init__(self, server, params):
+        import pymemcache.client.hash
+        import pymemcache.serde
+        super().__init__(server, params, library=pymemcache, value_not_found_exception=KeyError)
+        self._class = type('PyMemcacheClient', (PyMemcacheCompatMixin, self._lib.client.hash.HashClient), {})
+        self._options = {
+            'allow_unicode_keys': True,
+            'default_noreply': False,
+            'serde': pymemcache.serde.pickle_serde,
+            **self._options,
+        }
+
+    def touch(self, key, timeout=DEFAULT_TIMEOUT, version=None):
+        key = self.make_key(key, version=version)
+        return self._cache.touch(key, self.get_backend_timeout(timeout))
+
+
+class PyMemcacheCompatMixin:
+    """A mixin to work around API incompatibilities in pymemcache clients."""
+
+    def __init__(self, servers, *args, **kwargs):
+        servers = list(self._modify_servers(servers))
+        super().__init__(servers, *args, **kwargs)
+
+    @staticmethod
+    def _modify_servers(servers):
+        for server in servers:
+            if server.startswith('unix:'):
+                yield server[5:]
+            elif server.startswith('/'):
+                yield server
+            elif ':' not in server or server.endswith(']'):
+                yield (server, 11211)
+            else:
+                host, port = server.rsplit(':', 1)
+                yield (host, int(port))
+
+    def disconnect_all(self):
+        for client in self.clients.values():
+            client.close()
+
+    def get_many(self, keys, gets=False, *args, **kwargs):
+        # pymemcache.client.hash.HashClient may return {'key': False}
+        result = super().get_many(keys, gets, *args, **kwargs)
+        return {k: v for k, v in result.items() if v}
+
+    get_multi = get_many
```

Comment:
@Surgo I noticed that [lines similar to these](https://github.com/django-pymemcache/django-pymemcache/blob/b5a737e0507efac3765eaa9dc7a4346ee32e413c/djpymemcache/client.py#L46-L52) exist in `django-pymemcache`. It is unclear whether they are required as there are no test failures without them. Can you shed any light on why they may be required?

Reply:
Maybe pinterest/pymemcache#58 related alias. 
I think this alias is unnecessary.

CC: Original author @jezdez

Reply:
Currently, pymemcache has `get_multi`. 
So

```python
    def get_multi(self, keys, gets=False, *args, **kwargs):
        # pymemcache.client.hash.HashClient may return {'key': False}
        result = super().get_multi(keys, gets, *args, **kwargs)
        return {k: v for k, v in result.items() if v}
```

work too :)

Reply:
Ok. That makes sense. I'll also amend to only exclude the key if the value is `False` rather than falsy as we're trying to "undo" [this](https://github.com/pinterest/pymemcache/blob/54eb398eab4c625bf3696701f27e2a285477e1f5/pymemcache/client/hash.py#L375-L377). It also seems that this only happens if the client for a key cannot be found and [`ignore_exc` is set](https://github.com/pinterest/pymemcache/blob/54eb398eab4c625bf3696701f27e2a285477e1f5/pymemcache/client/hash.py#L155-L156), which is not the default.

Reply:
I've opened pinterest/pymemcache#291 to see if we can get the value changed from `False` as items explicitly stored with a value of `False` when `ignore_exc=True` will be removed too.